Newsletters: add 101 (2020-06-10)

harding opened the pull request:
- [x] Bitcoin Core PR review club section @jonatack 
- [x] `Eclair #1440` @adamjonas 
- [x] `LND #4251` @dongcarl

jonatack commented on _posts/en/newsletters/2020-06-10-newsletter.md:
Note: There are two places where the 3rd-person user is switched to 2nd-person "you" -- this paragraph, and the "Check hardware wallet compatibility" paragraph in Action Items.

harding replied:
I'm changing this paragraph to third person, but the Action Items section is definitely meant to refer directly to the reader.

jonatack commented on _posts/en/newsletters/2020-06-10-newsletter.md:
s/BIPs/BIP/

harding replied:
BIPs is "correct" here, at least per the convention I've been using in this section for more than a year (see `git grep '\[BIPs #'`).  I'm trying to convey that this is `BIPs [repository PR] #920` and not "BIP number 920" (which we would write as BIP920).

jonatack replied:
Oh right, the repo :hammer:

jonatack commented on img/posts/2020-06-fee-overpayment-attack.dot:
```diff
@@ -0,0 +1,56 @@
+  fake1_in_mine -> fake1_out_other;
+  fake1_in_other -> fake1_out_mine;
+
+  label = "Perceieved transaction 1\nCalculated fee: 0.0001 BTC";
```
"Perceived" here and line 34 below (and in the generated svg)

Great work on the diagram!

adamjonas commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,308 @@
+  transaction fee for a Bitcoin transaction is the difference between
+  the amounts of all the UTXOs that transaction spends and the amounts
+  of all the UTXOs it creates.  Transactions explicitly state the amounts
+  of the UTXOs they create but the amounts of the UTXOs they spend can
```
```suggestion
  of the UTXOs they create, but the amounts of the UTXOs they spend can
```

adamjonas commented on _posts/en/newsletters/2020-06-10-newsletter.md:
It is a dense sentence. Suggest breaking up.
```suggestion
    of all UTXOs being spent in that transaction. This makes the signature
```

harding replied:
With this proposed change, we'd have two sentences in a row that start with "This".  I'll make a slightly more invasive rephrasing to avoid that.

adamjonas commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```suggestion
  addresses generated from lexicographically sorted keys, such as those
```

adamjonas commented on _posts/en/newsletters/2020-06-10-newsletter.md:
increase 2x in the sentence
```suggestion
    using a separately-downloaded database to improve Bitcoin Core's
```

harding commented on _posts/en/newsletters/2020-06-10-newsletter.md:
Agree in principle with the nit, but I still feel that the lack of ability to send PSBTs from the GUI (without using the RPC console) warrants the additional modifier.

harding commented on _posts/en/newsletters/2020-06-10-newsletter.md:
Just checking that you meant so say "its" here (no trailing apostrophe)?

jonatack replied:
Indeed; quite the unfortunate typo. s/it's/its/

moneyball commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,311 @@
+    ![Fee overpayment attack illustration](/img/posts/2020-06-fee-overpayment-attack.dot.png)
+
+    Similar to the other attack described in Sander's 2017 email (see
+    [Newsletter #97][news97 spk commit]), this attack only affects
```
OOC what was newly revealed this week vs. Sander's 2017 email? Why wasn't this discussed more and addressed 3 years ago?

harding replied:
I don't know what "OOC" means, but as far as I know nothing new has been revealed this week.  I think it was discussed a bit more back then (e.g. on IRC) and this summary describes the attempts to resolve it at the protocol layer (e.g. Lau's SIGHASH2 proposal in 2018 and bip-taproot in 2019, both of which we summarized in Optech newsletters).

It appears from Trezor's description that they just learned about this attack when Rashid reported it and I'm guessing that that's also when other hardware wallets learned about it (or through Trezor's own disclosure, as it appears Rashid didn't contact Coinkite (ColdCard) or some other hardware wallet vendors).  I've been aware of this attack since shortly after Sander's published his description (I think I first heard about it on IRC) and it never occurred to me that it wasn't known to vendors.

I have a pet theory that Trezor didn't see Sanders's post because it occurred in a thread about PSBT and Trezor has historically been [opposed](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-August/013041.html) to hardware wallet interface standardization, Ledger has never been really involved in Bitcoin protocol development, and CoinKite hadn't started work on ColdCard at that point.

As to why nobody else worked on it, comments in #bitcoin-core-dev IRC indicate that nobody who knew about the attack for the past three years thought it was particularly urgent.  I'll update to sprinkle a few more links through the text.  My own personal take on it is that it isn't that severe of an issue but that probably all hardware wallet manufacturers will end up doing the same thing Trezor does just to prevent FUD from other HW manufacturers and to cover their asses against people claiming to lose money due to this vulnerability (whether legitimate or illegitimate---if you're a miner or know a trusted miner, it's easy to fake being a victim of this attack).

jonatack replied:
@harding Well said. This is my understanding as well, from the various IRC discussions of the past week.

jonatack replied:
At the last wallet meeting Friday, people weren't immediately convinced it was worth patching: http://www.erisian.com.au/bitcoin-core-dev/log-2020-06-05.html#l-346

moneyball commented on _posts/en/newsletters/2020-06-10-newsletter.md:
Should we also call out the software that is now broken such as BTCPay server and Wasabi if a user upgrades to the latest Trezor firmware? That is another drawback/tradeoff.

harding replied:
We do this, we just don't explictly name the software that's broken.

> This has broken compatibility with several wallets that interface with Trezor devices, either directly or through HWI.

harding commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,308 @@
+## News
+
+- **Time dilation attacks against LN:** Gleb Naumenko and Antoine Riard
+  [posted][naumenko td] to the Bitcoin-Dev mailing list a summary of a
```
How much do you hate this repetition?  I personally found it amusing that they summarized the summary of their paper and phrased this way to convey my amusement, but I'm happy to change it as you suggest if you still want that.

adamjonas replied:
I must defer to those that have a better sense of humor than I. Keep it.

harding commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,308 @@
+    through [HWI][topic hwi].  In cases where wallets have full copies
+    of the previous transactions or are able to obtain them, restoring
+    compatibility should just be a matter of updating the wallet code.
+    In other cases where wallets may not store copies of full previous
```
Changing to "full copies of previous transactions", which sounds better to me and matches the phrasing used in the previous sentence.

jnewbery commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,311 @@
+    an attacker could send two versions of the same transaction with two
+    segwit UTXOs to a hardware wallet.  In the first case, the attacker
+    would understate the amount of one of the UTXOs being spent; the
+    second case, they'd understate the amount of the second UTXO.
```
either "one of the UTXOs being spent ... the other UTXO being spent" or "the first UTXO being spent ... the second UTXO being spent"

harding replied:
It appears I pushed an update after you began reviewing this; I think that addressed your comment.  Could you please re-review?

jnewbery replied:
yes, this is now addressed.

jnewbery commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,311 @@
+    UTXOs, the hardware wallet would under-calculate the fee by that
+    same understated amount.  A user would authorize signing the first
+    transaction based on its understated fee.  The attacker would then
+    claim there was some minor problem and the user needs to sign the
```
nit: would claim _that_ there was .... and _that_ the user?

jnewbery commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,311 @@
+    transaction based on its understated fee.  The attacker would then
+    claim there was some minor problem and the user needs to sign the
+    same transaction again, but would instead get the user to sign the
+    second transaction (which would look identical to the user).
```
"would look identical to the user" is confusing since the transactions are not identical. How about "which would appear to be the same to the user since they pay the same outputs and same fee"

jnewbery commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,311 @@
+    same transaction again, but would instead get the user to sign the
+    second transaction (which would look identical to the user).
+    Although both transactions produced are individually invalid because
+    they each contain one signature that commits to a wrong amount, each
```
I'd remove the part about the invalid signature. The attacker doesn't need to provide a signature at all.

harding replied:
I don't understand this comment.

jnewbery replied:
Oh, I misunderstood. I didn't understand that both UTXOs are owned by the victim. Perhaps that needs to be explicitly stated for slow people like me?

harding commented:
Made suggested edits (thanks @jonatack and @adamjonas!), added some links to the recent #bitcoin-core-dev discussion about the fee overpayment attack, and added some additional info about its effect on PSBTs.

jnewbery commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,311 @@
+    benefits] and [hardware wallets][palatinus inpatient] believed would
+    allow signers to safely accept amounts from untrusted programs.
+
+    Unfortunately, signing a single UTXO's amount turned out to be
```
I'd emphasize that the 'attacker' is not able to directly steal any funds from the victim, simply make him burn bitcoin in fees.

jnewbery commented on _posts/en/newsletters/2020-06-10-newsletter.md:
Drop "perfectly". Claiming perfection is always dangerous!

jnewbery commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,324 @@
+[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
+BOLTs][bolts repo].*
+
+- [Eclair #1440][] allows for updating multiple channels via the API. Proposed
```
"allows for updating" sounds a little odd to me (the primary definition of "allow for" is to plan for something that might or might not happen (https://www.merriam-webster.com/dictionary/allow%20for). How about "adds support for updating" or "allows multiple channels to be updated"

ariard commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,329 @@
+    The paper describes an attack that can steal from lightweight
+    clients in as little as two hours.  For attacks against LN nodes
+    backed by a local Bitcoin full verification node, their shortest
+    duration attack requires eclipsing the victim for at least 24 hours.
```
In fact that's few hours, depending on implementation time-locks (see Table 1 : Default timelock (in blocks), timeout policy). 

You can also note, due to the lack of hardened eclipse protection in lightweight client and low-number of deployed lightweight servers, this attack is practical with high-chances of success _today_, assuming you have a channel with the victim.

ariard commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,329 @@
+    duration attack requires eclipsing the victim for at least 24 hours.
+    The authors note that the attacks probably apply to other
+    time-sensitive contract protocols and that the best solutions to the
+    problem all involve improving eclipse attack resistance.
```
The AltNet prroposal (https://github.com/bitcoin/bitcoin/pull/18988) has been quite motivated by this research. The current base layer p2p assumptions are limited with regards to LN security model.

ariard replied:
Blog post invite future research on inter-layer mapping or native-LN attacks. For more context, see IRC logs of LN dev meeting of today, which is about relevance of those topics.

harding commented:
Pushed two commits for the fee overpayment section.  One commit adds info about Ledger's response (and so makes some of the previous Trezor-focused text more generic); the other commit makes the changes suggested by @jnewbery.  The paragraph that contained the actual attack description is almost completely rewritten (and now two paragraphs), so re-review there would be appreciated.

@jnewbery's and @ariard's comments for other sections of the newsletter haven't been addressed yet.  I'm heading out to breakfast now, but I'll get to them within a few hours.

jonatack commented:
Review club writeup in #419.

jonatack commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,387 @@
+    spending segwit UTXOs without full copies of their previous
+    transactions.  Although requiring previous transactions does
+    maximize user safety against this attack, there are also arguments
+    for allowing hardware wallets to optionally continue to use segwit's
```
perhaps s/to use/using/ to align with "rather than breaking"

jonatack replied:
Bitcoin Core PR opened yesterday on this by Andrew Chow: https://github.com/bitcoin/bitcoin/pull/19215

As well as in HWI: https://github.com/bitcoin-core/HWI/pull/340

jonatack commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,387 @@
+    order to eliminate bugs, enhance security, and prepare for future
+    changes.  One forward-looking change that's attracted some attention
+    is the addition of an `asmap` configuration setting that allows
+    using a separately-downloaded database to improve Bitcoin Core's
```
nit: can omit hyphen

jonatack commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,387 @@
+- [LND #4251][] brings its REST interface's list of supported non-streaming
+  messages to parity with that of its gRPC interface. Previously, only a subset
+  of these messages (specifically those of the `Lightning` subserver) were
+  available via the REST interface. This is PR will be followed by [LND
```
s/is PR/PR/

jonatack commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,387 @@
+  print a warning when processing segwit inputs received from
+  non-updated software.  Other hardware wallets may release similar
+  updates in the future; contact the manufacturers for more information
+  about their plans.  See the *news* item below about the fee
```
perhaps s/news/News/ and s/item/section/

harding commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,387 @@
+    Ledger has made a similar change, although their [support
+    article][ledger update] says they only display a warning when
+    spending segwit UTXOs without full copies of their previous
+    transactions.  Although requiring previous transactions does
```
I think you missed a golden opportunity to start your comment by saying: "Although all those althoughs..."  :-)

jonatack commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,387 @@
+    ability to avoid connecting to too many peers all running on a
+    network owned by the same group, increasing its resistance to
+    [eclipse attacks][topic eclipse attacks] such as the [Erebus
+    attack][] (see [Newsletter #83][news83 asmap]).  However, one of the
```
Likely depends on who one talks to, but I'm not sure the terminology and description is ideal:

- The asmap feature and discussion specifically targeted Erebus: see https://github.com/bitcoin/bitcoin/pull/16702 and https://github.com/bitcoin/bitcoin/issues/16599.

- The researchers may be biased but do describe Erebus as a separate, third partitioning attack (see https://erebus-attack.comp.nus.edu.sg, section "How is the Erebus attack different from previous partitioning attacks?"): *There are two previous partitioning attacks, i.e., the Bitcoin hijacking (by Apostolaski et al., IEEE SP'17) and the Eclipse attack (by Heilman et al., USENIX Sec'15)*, followed by a table that details the distinguishing characteristics.

jonatack replied:
Perhaps the question is whether "eclipse attacks" has become a generic term for partitioning attacks, separate from the Eclipse attack (2015).

harding replied:
@jonatack Eclipse Attack is a term that predates Bitcoin.  Here's a paper from 2006: https://homepage.divms.uiowa.edu/~ghosh/eclipse.pdf .  I don't claim to be an expert, but I think eclipse attack is the generic name for isolating a node from any honest peer so that you can feed it misinformation.  Heilman's paper describe several *methods* for eclipsing a Bitcoin node, and the Erebus attack describes another method.  In short, I think the Erebus attack is a type of eclipse attack (and I previously updated the text to say that).

jonatack replied:
> @jonatack Eclipse Attack is a term that predates Bitcoin. Here's a paper from 2006

TIL - thanks! Will read.

harding commented:
Pushed edits for review feedback (thanks everyone!), plus a few small edits of mine to @jonatack's Bitcoin Core PR Review Club section.  ACK @adamjonas @dongcarl and @jonatack contributions (thanks!)

jonatack commented:
Thanks for the review club review/edits, @harding!

bitschmidty commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,393 @@
+[hw security]: https://en.bitcoin.it/wiki/Hardware_wallet#Security_risks
+[towns benefits]: https://bitcoincore.org/en/2016/01/26/segwit-benefits/#signing-of-input-values
+[palatinus inpatient]: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-October/013248.html
+[news11 sighash]: https://bitcoinops.org/en/newsletters/2018/09/04/#proposed-sighash-updates
```
s/
https://bitcoinops.org/en/newsletters/2018/09/04/#proposed-sighash-updates
/en/newsletters/2018/09/04/#proposed-sighash-updates

bitschmidty commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,393 @@
+  `sortedmulti` addition to the [output script descriptor][topic
+  descriptors] language that can make it easier to watch multisig
+  addresses generated from lexicographically sorted keys, such as those
+  used with ColdCard's native multisig support ([#68][news68
```
s/ColdCard/Coldcard

harding commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,387 @@
+  of the UTXOs they create, but the amounts of the UTXOs they spend can
+  only be found by looking at the previous transactions that created
+  those UTXOs.  The spending transaction simply commits to the txid and
+  location of the UTXOs it wants to spend, requiring other software that
```
Changing s/location/position/.  "Output index (vout)" is *exactly* the phrasing I'd use if I wanted to be precise, but in this case I just want to explain that the hardware wallet needs the full previous transaction (sans witness) in order to calculate the txid.

harding commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,393 @@
+    wallet to store that data plus require rescanning the block chain
+    for past wallet transactions.  Software that creates or updates
+    [Partially Signed Bitcoin Transactions][topic psbt] (PSBTs) will
+    need to be [upgraded][Bitcoin Core #12915] to include the full copies
```
Transposition :facepalm: Thanks!

jonatack commented:
ACK dec3b0b599a394a7763f35d6442974a040916a0b edits.

jnewbery commented on _posts/en/newsletters/2020-06-10-newsletter.md:
```diff
@@ -0,0 +1,393 @@
+    commits to an incorrect UTXO amount.  The compromised software then
+    claims that there was some minor problem that prevented the
+    transaction from being broadcast and tells the user they need to
+    re-sign the transaction, but the wallet doesn't re-send the first
```
s/wallet/software/. You've distinguished between 'hardware wallet' and 'compromised software' up to this point. Using the word wallet here is confusing.

jnewbery commented:
ACK 7e9828fdff1a66c37c9104e07ede66097e44a901 :weight_lifting:

bitschmidty commented:
Great newsletter, great teamwork!
✍️ @harding @adamjonas @dongcarl @jonatack 
🔎 @moneyball @jnewbery @ariard
